# 1.	Data Cleaning and Preparation

In [1]:
# import pandas to import dataset and data preprocessing
import pandas as pd

In [2]:
# loading all csv files into dataframe
df_sales = pd.read_csv('E:\Desktop_28062024\Guvi\Project2\DataSet\Sales.csv')
df_stores = pd.read_csv('E:\Desktop_28062024\Guvi\Project2\DataSet\Stores.csv')
df_products = pd.read_csv('E:\Desktop_28062024\Guvi\Project2\DataSet\Products.csv')
df_customers = pd.read_csv('E:\Desktop_28062024\Guvi\Project2\DataSet\Customers.csv')
df_exchange = pd.read_csv('E:\Desktop_28062024\Guvi\Project2\DataSet\Exchange_Rates.csv')

# Check datatype and null values of Sales dataframe 

In [18]:

df_sales.info()
df_sales.isnull().mean() * 100 
#df_sales.count()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62884 entries, 0 to 62883
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Order Number   62884 non-null  int64 
 1   Line Item      62884 non-null  int64 
 2   Order Date     62884 non-null  object
 3   Delivery Date  13165 non-null  object
 4   CustomerKey    62884 non-null  int64 
 5   StoreKey       62884 non-null  int64 
 6   ProductKey     62884 non-null  int64 
 7   Quantity       62884 non-null  int64 
 8   Currency Code  62884 non-null  object
dtypes: int64(6), object(3)
memory usage: 4.3+ MB


Order Number      0.000000
Line Item         0.000000
Order Date        0.000000
Delivery Date    79.064627
CustomerKey       0.000000
StoreKey          0.000000
ProductKey        0.000000
Quantity          0.000000
Currency Code     0.000000
dtype: float64

In [ ]:
# Delivery date feature is null for 75% of records.
# we can drop or ignore this null set of records for analysis.

In [ ]:
#df_sales = df_sales.drop(['Delivery Date'], axis=1)

# Convert date type for date fields
  -->2   Order Date     62884 non-null  object
  --> 3   Delivery Date  13165 non-null  object

In [3]:
df_sales['Order Date'] = pd.to_datetime(df_sales['Order Date'])

In [4]:
df_sales['Delivery Date'] = pd.to_datetime(df_sales['Delivery Date'])

# Feature(Column) name change as per DB standard

In [5]:
df_sales.rename(columns={'Order Number': 'Order_Number', 'Line Item': 'Line_Item', 'Order Date':'Order_Date','Delivery Date':'Delivery_Date','Currency Code':'Currency_Code'}, inplace=True)

# Check datatype and null values of store dataframe 

In [20]:
df_stores.info()
df_stores.isnull().mean() * 100 
#df_stores.count()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   StoreKey       67 non-null     int64  
 1   Country        67 non-null     object 
 2   State          67 non-null     object 
 3   Square Meters  66 non-null     float64
 4   Open Date      67 non-null     object 
dtypes: float64(1), int64(1), object(3)
memory usage: 2.7+ KB


StoreKey         0.000000
Country          0.000000
State            0.000000
Square Meters    1.492537
Open Date        0.000000
dtype: float64

In [ ]:
# 1% of record is null in store df for "SQUARE METER" feature.
# check and impute the data

In [22]:
df_stores[df_stores['Square Meters'].isnull()]

StoreKey Country   State  Square Meters Open Date
66         0  Online  Online            NaN  1/1/2010

In [ ]:
# only one records is null for Online so fill the value with zero

In [6]:
df_stores['Square Meters'].fillna(0, inplace=True)

In [25]:
df_stores.isnull().mean() * 100  # check for null values

StoreKey         0.0
Country          0.0
State            0.0
Square Meters    0.0
Open Date        0.0
dtype: float64

# Convert data type for date fields

In [7]:
df_stores['Open Date'] = pd.to_datetime(df_stores['Open Date'])

# Feature(Column) name change as per DB standard

In [8]:
df_stores.rename(columns={'Square Meters':'Square_Meters','Open Date':'Open_Date'},inplace=True)

# Check for null value in product dataframe

In [26]:
df_products.info()
df_products.isnull().mean() * 100 
#df_stores.count()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2517 entries, 0 to 2516
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ProductKey      2517 non-null   int64 
 1   Product Name    2517 non-null   object
 2   Brand           2517 non-null   object
 3   Color           2517 non-null   object
 4   Unit Cost USD   2517 non-null   object
 5   Unit Price USD  2517 non-null   object
 6   SubcategoryKey  2517 non-null   int64 
 7   Subcategory     2517 non-null   object
 8   CategoryKey     2517 non-null   int64 
 9   Category        2517 non-null   object
dtypes: int64(3), object(7)
memory usage: 196.8+ KB


ProductKey        0.0
Product Name      0.0
Brand             0.0
Color             0.0
Unit Cost USD     0.0
Unit Price USD    0.0
SubcategoryKey    0.0
Subcategory       0.0
CategoryKey       0.0
Category          0.0
dtype: float64

In [ ]:
# no null values in product dataset

# Feature(Column) name change as per DB standard

In [9]:
df_products.rename(columns={'Product Name':'Product_Name','Unit Cost USD':'Unit_Cost_USD','Unit Price USD':'Unit_Price_USD'},inplace=True)

# Check for null value in Customer dataframe

In [28]:
df_customers.info()
df_customers.isnull().mean() * 100 
#df_customers.count()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15266 entries, 0 to 15265
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CustomerKey  15266 non-null  int64 
 1   Gender       15266 non-null  object
 2   Name         15266 non-null  object
 3   City         15266 non-null  object
 4   State Code   15256 non-null  object
 5   State        15266 non-null  object
 6   Zip Code     15266 non-null  object
 7   Country      15266 non-null  object
 8   Continent    15266 non-null  object
 9   Birthday     15266 non-null  object
dtypes: int64(1), object(9)
memory usage: 1.2+ MB


CustomerKey    0.000000
Gender         0.000000
Name           0.000000
City           0.000000
State Code     0.065505
State          0.000000
Zip Code       0.000000
Country        0.000000
Continent      0.000000
Birthday       0.000000
dtype: float64

In [ ]:
# less then 1% of record is null in customer table "state code" is missing
# check and impute the data

In [10]:
df_customers[df_customers['State Code'].isnull()]['CustomerKey'].count()

10

In [ ]:
#10 records not having state code in customer table

# Convert data type for date fields

In [11]:
df_customers['Birthday'] = pd.to_datetime(df_customers['Birthday'])

# Feature(Column) name change as per DB standard

In [12]:
df_customers.rename(columns={'State Code':'State_Code','Zip Code':'Zip_Code'},inplace=True)

# Check for null value in Exchange dataframe

In [39]:
df_exchange.info()
df_exchange.isnull().mean() * 100 
#df_customers.count()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11215 entries, 0 to 11214
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Date      11215 non-null  object 
 1   Currency  11215 non-null  object 
 2   Exchange  11215 non-null  float64
dtypes: float64(1), object(2)
memory usage: 263.0+ KB


Date        0.0
Currency    0.0
Exchange    0.0
dtype: float64

# Convert data type for date fields

In [13]:
df_exchange['Date'] = pd.to_datetime(df_exchange['Date'])

# Merge dataframes based on the key combination

In [14]:
df_sales_cust = pd.merge(df_sales, df_customers, how='inner', on='CustomerKey')

In [15]:
df_sals_cust_prod = pd.merge(df_sales_cust, df_products, how='inner', on='ProductKey')

In [18]:
df_sals_cust_prod_store = pd.merge(df_sals_cust_prod, df_stores, how='inner', on='StoreKey')

In [21]:
df_all_ds = pd.merge(df_sals_cust_prod_store, df_exchange,how='inner' , left_on=['Order_Date', 'Currency_Code'], right_on=['Date', 'Currency'])

In [22]:
df_sales_dim = df_all_ds.copy()

In [23]:
df_sales_dim.isnull().sum()

Order_Number          0
Line_Item             0
Order_Date            0
Delivery_Date     49719
CustomerKey           0
StoreKey              0
ProductKey            0
Quantity              0
Currency_Code         0
Gender                0
Name                  0
City                  0
State_Code           30
State_x               0
Zip_Code              0
Country_x             0
Continent             0
Birthday              0
Product_Name          0
Brand                 0
Color                 0
Unit_Cost_USD         0
Unit_Price_USD        0
SubcategoryKey        0
Subcategory           0
CategoryKey           0
Category              0
Country_y             0
State_y               0
Square_Meters         0
Open_Date             0
Date                  0
Currency              0
Exchange              0
dtype: int64

In [24]:
df_sales_dim.describe()

Order_Number     Line_Item                     Order_Date  \
count  6.288400e+04  62884.000000                          62884   
mean   1.430905e+06      2.164207  2018-11-30 21:21:56.939126016   
min    3.660000e+05      1.000000            2016-01-01 00:00:00   
25%    1.121017e+06      1.000000            2018-01-25 00:00:00   
50%    1.498016e+06      2.000000            2019-02-06 00:00:00   
75%    1.788010e+06      3.000000            2019-11-23 00:00:00   
max    2.243032e+06      7.000000            2021-02-20 00:00:00   
std    4.532963e+05      1.365170                            NaN   

                       Delivery_Date   CustomerKey      StoreKey  \
count                          13165  6.288400e+04  62884.000000   
mean   2019-01-16 16:14:21.921762048  1.180797e+06     31.802144   
min              2016-01-06 00:00:00  3.010000e+02      0.000000   
25%              2018-05-05 00:00:00  6.808580e+05      8.000000   
50%              2019-03-05 00:00:00  1.261200e+06     37.000000   
75%              2019-12-18 00:00:00  1.686496e+06     53.000000   
max              2021-02-27 00:00:00  2.099937e+06     66.000000   
std                              NaN  5.859634e+05     22.978188   

         ProductKey      Quantity                       Birthday  \
count  62884.000000  62884.000000                          62884   
mean    1125.859344      3.144790  1968-12-05 20:00:44.882641056   
min        1.000000      1.000000            1935-02-04 00:00:00   
25%      437.000000      1.000000            1952-02-11 00:00:00   
50%     1358.000000      2.000000            1968-11-09 00:00:00   
75%     1650.000000      4.000000            1985-09-24 00:00:00   
max     2517.000000     10.000000            2002-01-30 00:00:00   
std      709.244010      2.256371                            NaN   

       SubcategoryKey   CategoryKey  Square_Meters  \
count    62884.000000  62884.000000   62884.000000   
mean       450.198238      4.469611    1260.640306   
min        101.000000      1.000000       0.000000   
25%        303.000000      3.000000     840.000000   
50%        501.000000      5.000000    1330.000000   
75%        602.000000      6.000000    2000.000000   
max        808.000000      8.000000    2105.000000   
std        212.710652      2.132682     767.739494   

                           Open_Date                           Date  \
count                          62884                          62884   
mean   2011-05-30 02:16:19.174352640  2018-11-30 21:21:56.939126016   
min              2005-03-04 00:00:00            2016-01-01 00:00:00   
25%              2010-01-01 00:00:00            2018-01-25 00:00:00   
50%              2010-01-01 00:00:00            2019-02-06 00:00:00   
75%              2013-06-07 00:00:00            2019-11-23 00:00:00   
max              2019-03-05 00:00:00            2021-02-20 00:00:00   
std                              NaN                            NaN   

           Exchange  
count  62884.000000  
mean       0.992533  
min        0.672500  
25%        0.894500  
50%        1.000000  
75%        1.000000  
max        1.665200  
std        0.161601

# Load data into SQL

In [ ]:
#Intalled mysqlclient, sqlalchemy to connect and load data into sql database

In [26]:
from sqlalchemy import create_engine

In [27]:
engine = create_engine('mysql://root:root@localhost/dt28')

In [29]:
df_sales.to_sql('sales', con=engine, if_exists='replace', index=False)
df_stores.to_sql('stores', con=engine, if_exists='replace', index=False)
df_products.to_sql('products', con=engine, if_exists='replace', index=False)
df_customers.to_sql('customers', con=engine, if_exists='replace', index=False)
df_exchange.to_sql('exchange', con=engine, if_exists='replace', index=False)

df_sales_dim.to_sql('sales_dim', con=engine, if_exists='replace', index=False)

62884

In [79]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15266 entries, 0 to 15265
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   CustomerKey  15266 non-null  int64         
 1   Gender       15266 non-null  object        
 2   Name         15266 non-null  object        
 3   City         15266 non-null  object        
 4   State_Code   15256 non-null  object        
 5   State        15266 non-null  object        
 6   Zip_Code     15266 non-null  object        
 7   Country      15266 non-null  object        
 8   Continent    15266 non-null  object        
 9   Birthday     15266 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(8)
memory usage: 1.2+ MB


In [80]:
df_customers.to_sql('customers', con=engine, if_exists='replace', index=False)

15266